xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/pvr/Timers.h

```diff
@@ -0,0 +1,873 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "../../AddonBase.h"
+#include "../../c-api/addon-instance/pvr/pvr_timers.h"
+
+//¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
+// "C++" Definitions group 6 - PVR timers
+#ifdef __cplusplus
+
+namespace kodi
+{
+namespace addon
+{
+
+//==============================================================================
+/// @defgroup cpp_kodi_addon_pvr_Defs_Timer_PVRTimerType_PVRTimerTypeIntValue class PVRTimerTypeIntValue
+/// @ingroup cpp_kodi_addon_pvr_Defs_Timer_PVRTimerType
+/// @brief **PVR add-on timer type value**\n
+/// Representation of a timer event related value.
+///
+//@{
+class PVRTimerTypeIntValue
```

Comment:
May I ask why this class does not follow the design pattern to wrap it's C counterpart (deriving it from `public CStructHdl<>`)?

Reply:
This and some others point not part for Kodi direct, them point to a struct inside another struct (here `PVR_TIMER_TYPE`), thought to hold there smaller by a standard C++ string store.

Look to add there too.

Reply:
Becomes replaced to (thought to hold smaller before :smile:):
```cpp
class PVRTimerTypeIntValue : public CStructHdl<PVRTimerTypeIntValue, PVR_ATTRIBUTE_INT_VALUE>
{
public:
  PVRTimerTypeIntValue() = default;
  PVRTimerTypeIntValue(int value, const std::string& description)
  {
    SetValue(value);
    SetDescription(description);
  }
  PVRTimerTypeIntValue(const PVRTimerTypeIntValue& data) : CStructHdl(data) {}
  PVRTimerTypeIntValue(const PVR_ATTRIBUTE_INT_VALUE* data) : CStructHdl(data) {}
  PVRTimerTypeIntValue(PVR_ATTRIBUTE_INT_VALUE* data) : CStructHdl(data) {}

  void SetValue(int value) { m_cStructure->iValue = value; }
  int GetValue() const { return m_cStructure->iValue; }
  void SetDescription(const std::string& description)
  {
    strncpy(m_cStructure->strDescription, description.c_str(),
            sizeof(m_cStructure->strDescription) - 1);
  }
  std::string GetDescription() const { return m_cStructure->strDescription; }
};
```

And the use place to:
```cpp
  void SetPreventDuplicateEpisodes(
      const std::vector<PVRTimerTypeIntValue>& preventDuplicateEpisodes,
      int preventDuplicateEpisodesDefault = -1)
  {
    m_cStructure->iPreventDuplicateEpisodesSize = preventDuplicateEpisodes.size();
    for (unsigned int i = 0; i < m_cStructure->iPreventDuplicateEpisodesSize &&
                             i < sizeof(m_cStructure->preventDuplicateEpisodes);
         ++i)
    {
      m_cStructure->preventDuplicateEpisodes[i].iValue = static_cast<const PVR_ATTRIBUTE_INT_VALUE*>(preventDuplicateEpisodes[i])->iValue;
      strncpy(m_cStructure->preventDuplicateEpisodes[i].strDescription, 
              static_cast<const PVR_ATTRIBUTE_INT_VALUE*>(preventDuplicateEpisodes[i])->strDescription,
              sizeof(m_cStructure->preventDuplicateEpisodes[i].strDescription) - 1);
    }
    if (preventDuplicateEpisodesDefault != -1)
      m_cStructure->iPreventDuplicateEpisodesDefault = preventDuplicateEpisodesDefault;
  }

  /// @brief To get with @ref SetPreventDuplicateEpisodes changed values.
  std::vector<PVRTimerTypeIntValue> GetPreventDuplicateEpisodes() const
  {
    std::vector<PVRTimerTypeIntValue> ret;
    for (unsigned int i = 0; i < m_cStructure->iPreventDuplicateEpisodesSize; ++i)
      ret.emplace_back(m_cStructure->preventDuplicateEpisodes[i].iValue,
                       m_cStructure->preventDuplicateEpisodes[i].strDescription);
    return ret;
  }
```